Common/Data/UniverseSelection/ConstituentsUniverseData.cs

```diff
@@ -0,0 +1,96 @@
+    /// Custom base data class used for <see cref="ConstituentsUniverse"/>
+    /// </summary>
+    public class ConstituentsUniverseData : BaseData
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CoarseFundamental"/> class
+        /// </summary>
+        public ConstituentsUniverseData()
+        {
+            DataType = MarketDataType.Auxiliary;
+        }
+
+        /// <summary>
+        /// The end time of this data.
+        /// </summary>
+        public override DateTime EndTime
+        {
+            get { return Time + QuantConnect.Time.OneDay; }
+            set { Time = value - QuantConnect.Time.OneDay; }
+        }
+
+        /// <summary>
+        /// Return the URL string source of the file. This will be converted to a stream
+        /// </summary>
+        /// <param name="config">Configuration object</param>
+        /// <param name="date">Date of this source file</param>
+        /// <param name="isLiveMode">true if we're in live mode, false for backtesting mode</param>
+        /// <returns>String URL of source file.</returns>
+        public override SubscriptionDataSource GetSource(SubscriptionDataConfig config, DateTime date, bool isLiveMode)
+        {
+            var universe = config.MappedSymbol.Substring(config.MappedSymbol.LastIndexOf('-') + 1).ToLower();
+
+            if (isLiveMode)
+            {
+                date = date.AddDays(-1);
+            }
+            var path = Path.Combine(Globals.DataFolder,
+                config.SecurityType.SecurityTypeToLower(),
+                config.Market,
+                "universes",
+                config.Resolution.ResolutionToLower(),
+                universe,
+                $"{date:yyyyMMdd}.csv");
+            return new SubscriptionDataSource(path, SubscriptionTransportMedium.LocalFile, FileFormat.Csv);
+        }
+
+        /// <summary>
+        /// Reader converts each line of the data source into BaseData objects. Each data type creates its own factory method, and returns a new instance of the object
+        /// each time it is called.
+        /// </summary>
+        /// <param name="config">Subscription data config setup object</param>
+        /// <param name="line">Line of the source document</param>
+        /// <param name="date">Date of the requested data</param>
+        /// <param name="isLiveMode">true if we're in live mode, false for backtesting mode</param>
+        /// <returns>Instance of the T:BaseData object generated by this line of the CSV</returns>
+        public override BaseData Reader(SubscriptionDataConfig config, string line, DateTime date, bool isLiveMode)
+        {
+            try
+            {
+                var csv = line.Split(',');
+                var preselected = new ConstituentsUniverseData
+                {
+                    Symbol = new Symbol(SecurityIdentifier.Parse(csv[0]), csv[1]),
+                    Time = isLiveMode ? date.AddDays(-1) : date
```

Comment:
interesting discrepancy between live and backtesting -- I don't see something similar in `CoarseFundamental` which makes me nervous. Is this `date.AddDays(-1)` really highlighting a bug in the data generator algorithm which is saving to the incorrect date?

Reply:
Something just seems off about this, but if you're confident that things are consistent that than works for me. Also, this would _not_ produce a bug, as the application of these dates are completely within the silo of universe data being generated -- meaning -- so long as the type that generates the data and the type that reads the data agree on some 'convention' for specifying the time/end time, then we'll get the right result.  The only concern is we might have discrepancies between various 'silos', such as `CoarseFundamental` vs this constitutent data.